yarn-project/noir-contracts/src/contracts/lending_contract/src/main.nr

```diff
@@ -0,0 +1,323 @@
+    }
+
+    open internal fn _borrow(
+        inputs: PublicContextInputs,
+        owner: Field,
+        amount: Field
+    ) -> Field {
+        // Access control.
+        let _void = call_public_function_no_args(inputs.call_context.storage_contract_address, 1259373467)[0];
+
+        let storage = Storage::init();
+
+        let coll_loc = storage.collateral.at(owner);
+        let collateral: Field = coll_loc.read();
+
+        let debt_loc = storage.static_debt.at(owner);
+        let static_debt: Field = debt_loc.read();
+
+        assert (static_debt as u120 + amount as u120 < collateral as u120);
+
+        debt_loc.write(static_debt + amount);
+
+        // @todo @LHerskind Transferring funds to the recipient.
+
+        1
+    }
+
+    // This don't need to be on behalf of self. We should be able to repay on behalf of someone else.
+    fn repay_private(
+        inputs: PrivateContextInputs,
+        secret: Field,
+        owner: Field,
+        amount: Field
+    ) -> distinct pub abi::PrivateCircuitPublicInputs {
+        let mut initialContext = Context::new(inputs, abi::hash_args([
+            secret,
+            owner,
+            amount
+        ]));
+
+        assert (!((secret == 0) as bool & (owner == 0) as bool));
+        let mut account = owner;
+        let me = inputs.call_context.msg_sender;
+
+        if (secret != 0) {
+            account = Account::new(me, secret).key();
+        }
+
+        // @todo @lherskind Transfer tokens from me to this contract.
+
+        let (_callStackItem, mut context) = PublicCallStackItem::call(
+            inputs.call_context.storage_contract_address, 
+            3985016136,
+            [account, amount], 
+            initialContext
+        );
+
+        context.finish()
+    }
+
+    open fn repay_public(
+        inputs: PublicContextInputs,
+        owner: Field,
+        amount: Field
+    ) -> Field {
+        // @todo @LHerskind Transfer tokens from msg.sender to this contract.
+        call_public_function(inputs.call_context.storage_contract_address, 3985016136, [owner, amount])[0]
+    }
+
+    open internal fn _repay(
+        inputs: PublicContextInputs,
+        owner: Field,
+        amount: Field
+    ) -> Field {
+        let _void = call_public_function_no_args(inputs.call_context.storage_contract_address, 1259373467)[0];
+
+        let storage = Storage::init();
+        // Should constrain access from private.
+        // should only be callable from self from private.
+
+        let debt_loc = storage.static_debt.at(owner);
+
+        // @todo @LHerskind using "as" after using maps key => field have issues.
+        // Need to be explicitly specified as field first.
+        let static_debt: Field = debt_loc.read();
+
+        assert (static_debt as u120 >= amount as u120);
+
+        debt_loc.write(static_debt - amount);
+        1
```

Comment:
do we need the return type here? i dont think pub needs a return value anymore

Reply:
If there is no return on the public functions, there will be issues as it returns `[null]`, think it is because it is using the length for the return values or so, so it don't match the abi types and screams. @sirasistant wdyt is the best solution here, just having a return value or look at supporting a null

Reply:
Hmmm I think this was the same reason that we were using padding in the past, since the sequencer doesn't have access to the ABI it doesn't know how many return values to extract from the witness, so we force to always return a fixed amount (one). I'd say for now just return a constant field and I'll take a look at doing the same with returns as I did with args